### PR TITLE
Spacer: Show tooltip with height value on resize

### PR DESCRIPTION
**Pull request by ItsJonQ**
## Description

Screen Capture on 2020-06-10 at 16-38-00

This update enhances the **Spacer** block by rendering a tooltip element above the cursor when the user engages in a "drag to resize" interaction.

#### ✨ New `<ResizeTooltip />` component

The tooltip comes from a new `<ResizeTooltip />` component. This component is a general resize-aware component that renders a tooltip on size changes. It isn't bound to `<ResizableBox />` (which is what Spacer uses), but it can be integrated with it.

As an aside... This component has 2 different positions:

* `bottom`
* `corner` (default)

##### 📄 Corner

Screen Capture on 2020-06-10 at 16-06-28

The Tooltip will render on the top/right of the parent element.

##### ⬇️ Bottom

Screen Shot 2020-06-11 at 10 03 13 AM

The Tooltip will render on the bottom/center of the parent element.

#### 🎨 Design

As far as the Tooltip design goes... I went with a UI that was in the spirit of "G2". Something minimal, sporting the all black/white look.

## How has this been tested?

* Tested in local Storybook
* Tested in local Gutenberg

To test...

* Run `npm install`
* Run `npm run dev`
* Add a Spacer block
* Drag to resize. You should see the Tooltip.
* Drag the RangeSlider (sidebar) to resize the Spacer. You should NOT see the Tooltip.

## Types of changes

* Add `react-resize-aware` as dependency in `@wordpress/components` (it's also used in `@wrodpress/compose`)
* Add new `<ResizeTooltip />` component
* Exports `ResizeTooltip` from `@wordpress/components`
* Integrate `<ResizeTooltip />` into Spacer block

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. 

The idea for this PR originally came from:
https://github.com/WordPress/gutenberg/issues/22956

### CODE REVIEW

**Comment by ItsJonQ** [comment]
@MichaelArestad Thank you!! I pushed a fix. Try now 😊 

(Really appreciate the feedback!)

File: packages/components/src/resize-tooltip/styles/resize-tooltip.styles.js
Comment (mcsf) [inline packages/components/src/resize-tooltip/styles/resize-tooltip.styles.js:35]: We should indent these lines to make it clear they are part of the condition.

File: packages/components/src/resize-tooltip/utils.js
Comment (mcsf) [inline packages/components/src/resize-tooltip/utils.js:149]: Just looking at the code, I don't think this cleanup ever does anything. That's because the `handle*` functions are re-instantiated every time the effect runs.

The other thing is that I believe we could be more judicious about the dependencies of this effect. By refactoring the handler functions and properly looking at what values are needed, I believe we could reduce the `[ width, height, isActive, isDragging ]` dependency list, and therefore not tear down and reset this effect so often.

File: packages/components/src/resize-tooltip/index.js
Comment (mcsf) [reply packages/components/src/resize-tooltip/index.js:35]: By using one piece of state composed of an object like this, this component can't skip effects/rendering whenever the state changes but the coordinates themselves are unchanged. This affects other components affected by this shape, like Label.

It's something to think about and compare for performance against having two props `cursorX` + `cursorY`.

File: packages/components/src/resize-tooltip/styles/resize-tooltip.styles.js
Comment (ItsJonQ) [reply packages/components/src/resize-tooltip/styles/resize-tooltip.styles.js:35]: This is how my VSCode auto-formatted it (with Prettier and ESLint).
I admit... it doesn't look very nice. I'll rewrite it another way 🙌 

Thanks for the note!

File: packages/components/src/resize-tooltip/utils.js
Comment (ItsJonQ) [reply packages/components/src/resize-tooltip/utils.js:181]: Thank you for pointing this out! I just pushed up some changes which should feel better (from a code/useEffect perspective). Let me know what you think of these revisions 🙏

File: packages/components/src/resize-tooltip/index.js
Comment (ItsJonQ) [reply packages/components/src/resize-tooltip/index.js]: Great suggestion! Done :)

**Comment by MichaelArestad** [comment]
This is much better! I'm not seeing either of those bugs anymore. Well done!

**Comment by ItsJonQ** [comment]
@MichaelArestad Thank you!!

@mcsf 's feedback helped me rethink + optimize things 😉

File: packages/components/src/resize-tooltip/utils.js
Comment (mcsf) [inline packages/components/src/resize-tooltip/utils.js:52]: Typo at `Awere`.

File: packages/components/src/resize-tooltip/utils.js
Comment (mcsf) [inline packages/components/src/resize-tooltip/utils.js:51]: Our situation in WP-JS is a little awkward when it comes to multi-line comments, but in any case [we should avoid the leading double asterisk](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/#multi-line-comments) unless it's opening a JSDoc block.

File: packages/components/src/resize-tooltip/utils.js
Comment (mcsf) [reply packages/components/src/resize-tooltip/utils.js:181]: > Just looking at the code, I don't think this cleanup ever does anything. That's because the `handle*` functions are re-instantiated every time the effect runs.

I don't why I said this, as it didn't make sense: the registered listeners are properly referenced in the tear-down function returned by the effect, so we should always be able to remove the old event listeners.

Comment (mcsf) [reply packages/components/src/resize-tooltip/utils.js:181]: > Let me know what you think of these revisions

So I've been going back and forth looking at this effect and feeling generally unsatisfied when pondering things like the dependencies, and I think I know why. We're dealing with effects of first degree and effects of second degree, so to speak: the first-degree effects are the registration of the `mouse*` event listeners; the second-degree ones are the listeners themselves. This makes the reasoning about dependencies more confusing.

If we were stricter about constructing these effects, we might refactor this as follows:

```js
export function useResizeLabel( … ) {
  // ...

  const clearMoveTimeout = useCallback( () => {
    if ( moveTimeoutRef.current ) {
      clearTimeout( moveTimeoutRef.current );
    }
  }, [ moveTimeoutRef ] );

  const handleOnMouseDown = useCallback( () => {
    if ( ! isCursor ) return;

    setIsDragging( true );
    clearMoveTimeout();
    unsetMoveXY();
  }, [
    // meaningful dependencies (mutable)
    isCursor,
    // formal dependencies (assumed constant)
    setIsDragging, clearMoveTimeout, unsetMoveYX,
  ] );

  const handleOnMouseMove = useCallback( /* same as above */ );

  const handleOnMouseUp = useCallback( /* same as above */ );
  
  useEffect( () => {
    document.addEventListener( 'mousedown', handleOnMouseDown );
    return () => {
      document.removeEventListener( 'mousedown', handleOnMouseDown );
    };
  }, [ handleOnMouseDown ] );

  useEffect( /* same as above for `mousemove` */ );

  useEffect( /* same as above for `mouseup` */ );

  return { ... };
}
```

Note those comments about "meaningful" vs. "formal" dependencies. That was my of not making a decision as to whether we should be a strict as possible when listing dependencies or not. :) I'll leave that up to you.

**Comment by shaunandrews** [comment]
Love this. I do wonder about the capitalization of "PX". Maybe its just the size of that unit's label, but I wonder if lowercase ("px") would be better? I guess its mostly personal preference.

File: packages/components/src/resize-tooltip/utils.js
Comment (ItsJonQ) [reply packages/components/src/resize-tooltip/utils.js]: Oop! 🙊 . Thank you for the reminder! This is something I have to get use to!

**Comment by ItsJonQ** [comment]
@shaunandrews  + @MichaelArestad 

Lowecase `px`

Screen Shot 2020-06-11 at 2 37 24 PM

Screen Capture on 2020-06-11 at 14-37-19

Vote with 👍 / 👎 

Thank you!

File: packages/components/src/resize-tooltip/utils.js
Comment (ItsJonQ) [reply packages/components/src/resize-tooltip/utils.js:181]: Thanks for the detailed feedback @mcsf !

I'm still unsure of when `useEffect`/`useCallback` dependencies should be used. I use to be in the camp of over-using `useCallback` (and would have done what you had suggested).

However, after reading Kent C Dodd's posts on `useCallback`, I've eased up a lot:
https://kentcdodds.com/blog/usememo-and-usecallback

Nowadays, I'm more in the mindset of...
Use `useCallback` if it causes an obvious problem (too many executions, infinite loops, slow perf). Otherwise, it should be okay.

That being said! I'm happy to make changes to accommodate the preferred styles of the project 🙇

**Comment by ItsJonQ** [comment]
Thanks @MichaelArestad ! Lowercase it is :)

**Comment by shaunandrews** [comment]
Do we use units like this elsewhere? What does padding do?

File: packages/components/src/resize-tooltip/utils.js
Comment (mcsf) [reply packages/components/src/resize-tooltip/utils.js:181]: > I'm happy to make changes to accommodate the preferred styles of the project

First, a disclaimer: this is really not a preferred style, just the result of my spitballing, and it's important to remember that in general, when it comes to React hooks, much is up in the air! I don't think you should just change to what I suggest. :)

I totally agree on being careful about `useCallback`, and in that aspect my example is really not compelling.

My focus was on isolating each effect as much as possible and then being _exhaustive_ about its dependencies. I use that term based on the idea of the [`exhaustive-deps` lint rule](https://github.com/facebook/react/issues/14920). The idea is quite interesting in theory.

In my mind, it can prevent a lot of subtle bugs that could happen, especially in the long term as components are tweaked and it becomes harder to trace the implicit (i.e. not exhaustively declared) dependencies of an effect or other hook.

In practice, though, it becomes awkward. Any function called inside an effect or callback is technically a dependency, and if we're dealing with a function declared in the body of the component then this dependency changes all the time. At that point, we have three choices:

1. Drop the requirement for "exhaustive" deps
2. Generously use `useCallback` to cache those functions so that the effect in question doesn't re-run for no reason
3. Accept that the effect will run frequently

This obviously only opens the door for more questions. :) Like, for choices 2 and 3, how much is too much function caching, and how much it too much cache busting? For choice 1, how can design hooks so as to minimise bugs related to state, cache and closure?

* * *

Finally, and to help unblock things here: I really don't mind that we write effects in the way that makes most intuitive sense and is performant. In the case of these effects, it's clear that the the point is to register three DOM-event handlers that will more or less be active for as long as the component exists. We can totally keep the effects as you wrote them. ;)

**Review by mcsf — COMMENTED** [review]
Looking at `ResizeTooltop`, what if we didn't have `cursor` as a possible `position`? Might not we get by with either `bottom` or `corner`?

I ask for a couple of reasons. I think there is a difficult balance to find between working "one need at a time", and building components that aim to be framework-ready (customisable, reusable) from the start. My personal impression from this PR is that we may be leaning too much towards the latter. I'd like to start off with _less_: fewer modes, fewer overrides* (e.g. do we need to pass `zIndex`? maybe we do, but I'd like to know the reasoning).

The other thing is that it would be good to have a baseline — in terms of UX, of performance, of any such criterion — against which to measure any improvements. In this case, we would err on the side of simplicity, both in the complexity of the implementation and in the richness of the interactions, by only adding one or two of the tool-tip positioning modes. That way, any subsequent PR to improve or expand would give us much more insight on what works best.

*: I always worry about options and overrides because they become interfaces to mind when thinking about backwards compatibility in the present and in the future.

**Comment by ItsJonQ** [comment]
@mcsf Those are fair points. I've removed the "cursor" position from the ResizeTooltip. I've adjusted some logic and styles to improve the integration of `bottom` position.

We can revisit adding a cursor positioned Tooltip in the future, if needed 👍

**Comment by mcsf** [comment]
> @mcsf Those are fair points. I've removed the "cursor" position from the ResizeTooltip. I've adjusted some logic and styles to improve the integration of `bottom` position.
> 
> We can revisit adding a cursor positioned Tooltip in the future, if needed 👍

Thanks for humouring me!

**Comment by ItsJonQ** [comment]
> Thanks for humouring me!

Anytime 😉 ❤️

File: packages/block-library/src/spacer/edit.js
Comment (mcsf) [inline packages/block-library/src/spacer/edit.js:85]: Have you considered rendering actually _in_ ResizableBox? If there are reasons to keep it separate, do let me know. Otherwise, it might be a boon in developer experience that ResizableBox could handle everything related to resizing on its own.

I can't help but think that `axis` and `position` can be derived from ResizableBox's `enable` and to some extent `size` props. If we were to move Tooltip into Box the consumer of the component wouldn't have to worry about such extra props.

**Review by mcsf — COMMENTED** [review]
Everything is very detailed and carefully laid out, and I appreciate that.

I did have a question about how to fit ResizeTooltop in ResizableBox, though (the position I tend to come from is to minimise how much is exposed at first, and see over time what should be opened up).

File: packages/block-library/src/spacer/edit.js
Comment (ItsJonQ) [reply packages/block-library/src/spacer/edit.js:85]: @mcsf I have considered putting it in `ResizableBox` :). That's how this idea started out.
The reason I thought of keeping it independent was for added flexibility. That is, to show dimensions on resize that aren't triggered by `ResizableBox`. For example, browser `window` resize.

To start, we could have this be a default (experimental) feature of `ResizableBox`. Something like `<ResizableBox __experimentalShowTooltip={true} />` would show the Tooltip.

What do you think? :)

Comment (mcsf) [reply packages/block-library/src/spacer/edit.js:85]: >  The reason I thought of keeping it independent was for added flexibility. That is, to show dimensions on resize that aren't triggered by `ResizableBox`. For example, browser `window` resize.

Hm, I could see that as an interaction, although personally I would wonder _when_ we actually expect users to resize their proper window in a context where we'd show the tool tip. :)

So I think we're still in this territory: "the position I tend to come from is to minimise how much is exposed at first, and see over time what should be opened up". Internally, ResizeTooltip is a good component to build separately, but I would **keep it private** to our components.

> To start, we could have this be a default (experimental) feature of `ResizableBox`. Something like `<ResizableBox __experimentalShowTooltip={true} />` would show the Tooltip.

I'd be happy to see that experiment. :) Don't feel like you need to do it in this PR, though, as long as we keep the right things private/experimental.

Comment (ItsJonQ) [reply packages/block-library/src/spacer/edit.js:85]: > Internally, ResizeTooltip is a good component to build separately, but I would keep it private to our components.

@mcsf Sounds good to me 👍 . With this PR, in it's current state, does exporting it as `__experimental` qualify? Or would it be better to integrate it into `ResizableBox` and not export it at all?

Thanks!

Comment (mcsf) [reply packages/block-library/src/spacer/edit.js:85]: > in it's current state, does exporting it as `__experimental` qualify? Or would it be better to integrate it into `ResizableBox` and not export it at all?

If you don't mind too much, integrating in `ResizableBox` and not exporting would be the best, IMO!

Comment (ItsJonQ) [reply packages/block-library/src/spacer/edit.js:85]: @mcsf Alrighty! Let's go with that for now :)

**Comment by ItsJonQ** [comment]
@mcsf I just moved the ResizeTooltip under ResizableBox. Lemme know if this works! <3

File: docs/manifest.json
Comment (mcsf) [inline docs/manifest.json:1025]: Should no longer be in the manifest, right?

**Review by mcsf — APPROVED** [review]
Assuming the manifest thing is handled, LGTM! Nice work, and thanks for the ping-ponging!

File: docs/manifest.json
Comment (ItsJonQ) [reply docs/manifest.json:1025]: I'm... not sure. This was auto generated. There's a `README.md` inside the directory. I guess I can remove it? Ideally I'd like to preserve the docs though.

An alternative would be to inline comment some of the documentation (for now)

Comment (mcsf) [reply docs/manifest.json:1025]: Hmm... yeah, I see now. Well, I don’t have a preference. Maybe @nosolosw has a different take on how we should document components and let these tools work for us? Anyway, LGTM :)

**Comment by ItsJonQ** [comment]
Alrighty! Travis is green and happy. Thanks for the review @mcsf !!

I'll merge it up

**Comment by mcsf** [comment]
🎉

File: docs/manifest.json
Comment (nosolosw) [reply docs/manifest.json:1025]: I took a peek at how the script that processes the handbook docs works. It seems that it looks for README files within the components folder ([see](https://github.com/WordPress/gutenberg/blob/master/docs/tool/manifest.js#L9)) and then updates the manifest used to expose them in the handbook (see [ResizableBox page](https://developer.wordpress.org/block-editor/components/resize-tooltip/)).

I'm unfamiliar with how we deal with subcomponents that aren't exposed via the `@wordpress/components` package. I had a quick look and this seems to be the only case? Anyway, it seems that we should either expose the resize-tooltip as a component proper or look at ways to remove its page from the handbook as it can't be used by external consumers of the API. For the latter, we could go with a quick hack (use a lowercased readme, use any other name for the file) or fix the issue (update the glob pattern to only match README files within the top-level component folder).

**Pull request merged**
